### PR TITLE
In-memory datastore for aleph peers

### PR DESCRIPTION
**Pull request by yusefnapora**
This adds a `Datastore` class that's backed by the in-memory [MemDOWN](https://github.com/level/memdown) backend for [LevelUP](https://github.com/Level/levelup), the node leveldb library.

I went this route rather than rolling my own in-memory kv store because

- this gives us an easy path to a persistent datastore by using the leveldb (or rocksdb) backend
- it forced me to use an async put/get interface, which we'll need if we do the above
- it was pretty easy

Aleph nodes now supports the `/mediachain/data` protocol to fetch data, but writing to the datastore needs to happen in-process (no push or merge support yet).  Still, merge should be fairly easy to accomplish, since we've got remote query working already.

I added a simple test between two aleph nodes, but haven't done an integration test yet.

### CODE REVIEW

**Comment by vyzo** [comment]
:+1: for having a real datastore.

Basically, in terms of workflow, aleph can be the workspace where you are building your data, and when you are ready to publish you push them to your concat.

File: src/peer/datastore.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+const DefaultOptions: DatastoreOptions = {
+  backend: 'memory',
+  location: '/aleph/data'
+}
+
+class Datastore {
+  db: Levelup
+
+  constructor (options: ?DatastoreOptions = null) {
+    if (options == null) {
```
Comment (parkan) [inline src/peer/datastore.js]: wouldn't it be easier to have `options: ?DatastoreOptions = {}` and merge it in regardless of are you trying to optimize?

Comment (yusefnapora) [inline src/peer/datastore.js]: yeah, it definitely would... but flow gives an error because `{}` isn't a valid value for `DatastoreOptions`

File: src/peer/node.js [hunk 2]
```diff
@@ -270,8 +278,57 @@ class MediachainNode {
...
+
+      // make sure we close the stream when the node stops
+      this.p2p.newAbortable(),
+
+      // convert each request into a stream of keys, followed by a StreamEnd message object
+      pull.map(req => [...req.keys, {end: {}}]),
+      pull.flatten(),
+
+      // fetch values from the datastore, returning them in the same order as the input
+      paramap((keyOrEnd: string | Object, callback) => {
```
Comment (parkan) [inline src/peer/node.js]: hmm I wonder if we should have protocol compliance tests to make sure aleph and concat behaviors do not diverge in corner cases

File: src/peer/datastore.js (hunk 1 above)
Comment (parkan) [inline src/peer/datastore.js]: make the members optional? I'm actually not sure what `locaiton` even means for `memory` backend

File: src/peer/node.js (hunk 2 above)
Comment (yusefnapora) [inline src/peer/node.js]: a protocol test suite would be awesome.  maybe lets try to get that together after the break?
